crates/red_knot_python_semantic/src/types/call/bind.rs

```diff
@@ -722,6 +751,72 @@ impl<'db> Bindings<'db> {
                         }
                     }
 
+                    Some(KnownClass::TypeVar) => {
+                        let Some(containing_assignment) = containing_assignment else {
+                            // TODO: Raise a diagnostic if TypeVar is called without being
+                            // immediately assigned to local variable.
+                            continue;
+                        };
+
+                        let [Some(_name), constraints, bound, default, _contravariant, _covariant, _infer_variance] =
+                            overload.parameter_types()
+                        else {
+                            continue;
+                        };
+
+                        // TypeVar can only be called when the result is immediately assigned to a
+                        // variable. Reuse the name that is being assigned to.
+                        // TODO: Raise a diagnostic if the name parameter doesn't match the name
+                        // being assigned to.
+                        let name = match containing_assignment.kind(db) {
+                            DefinitionKind::Assignment(assignment) => match assignment.target() {
+                                ast::Expr::Name(name) => name.id.clone(),
+                                _ => continue,
+                            },
+                            _ => continue,
+                        };
```

Comment:
I haven't checked this fully but this function is called indirectly from `Type::try_call`, meaning we shouldn't access any AST nodes in it. This might be the cause for the significant incremental perf regression

Reply:
Thanks for catching that!  I can try to move the check that needs this back up into `infer.rs`

Reply:
Done! Doing that also let me actually add the diagnostics for invalid legacy typevar construction

Reply:
(Note that we still pass the `containing_assignment` in as a parameter, since that's the definition we need to record for a legacy typevar. We just don't access its `kind` field here)

Reply:
Are these changes local because line 771 still access the `kind` field?

Reply:
I'm not sure whether I did "submit comment" or "push" first, but the changes should be up now if I messed up the order earlier

Reply:
Unfortunately this did not fix the performance regression

Reply:
Yeah, unfortunately not.